README
NAME

    NewRelic::Agent - Perl Agent for NewRelic APM

VERSION

    version 0.0532

SYNOPSIS

        use NewRelic::Agent;
    
        my $agent = NewRelic:Agent->new(
            license_key => 'abc123',
            app_name    => 'REST API',
        );
    
        $agent->embed_collector;
        $agent->init;
        my $txn_id = $agent->begin_transaction;
        ...
        my $err_id = $agent->end_transaction($txn_id);

DESCRIPTION

    This module provides bindings for the NewRelic
    <https://docs.newrelic.com/docs/agents/agent-sdk/getting-started/new-relic-agent-sdk>
    Agent SDK.

METHODS

 new

    Instantiates a new NewRelic::Agent client object.

        my $agent = NewRelic::Agent->new(
            license_key          => $license_key,
            app_name             => $app_name,
            app_language         => $app_language,         #optional
            app_language_version => $app_language_version, #optional
        );

    Parameters

    - license_key

      A valid NewRelic license key for your account.

      This value is also automatically sourced from the
      NEWRELIC_LICENSE_KEY environment variable.

    - app_name

      The name of your application.

      This value is also automatically sourced from the NEWRELIC_APP_NAME
      environment variable.

    - app_language

      The language that your application is written in.

      This value defaults to perl, and can also be automatically sourced
      from the NEWRELIC_APP_LANGUAGE environment variable.

    - app_language_version

      The version of the language that your application is written in.

      This value defaults to your perl version, and can also be
      automatically sourced from the NEWRELIC_APP_LANGUAGE_VERSION
      environment variable.

 embed_collector

    Embeds the collector agent for harvesting NewRelic data. This should be
    called before init, if the agent is being used in Embedded mode and not
    Daemon mode.

    Example:

        $agent->embed_collector;

 init

    Initialize the connection to NewRelic.

    Example:

        $agent->init;

 begin_transaction

    Identifies the beginning of a transaction, which is a timed operation
    consisting of multiple segments. By default, transaction type is set to
    WebTransaction and transaction category is set to Uri.

    Returns the transaction's ID on success, else negative warning code or
    error code.

    Example:

        my $txn_id = $agent->begin_transaction;

 set_transaction_name

    Sets the transaction's name.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_name($txn_id, 'Create Account');

 set_transaction_request_url

    Sets the transaction's request url. The query part of the url is
    automatically stripped from the url.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_request_url($txn_id, 'api.myapp.com/users/123');

 set_transaction_max_trace_segments

    Sets the maximum number of trace segments allowed in a transaction
    trace. By default, the maximum is set to 2000, which means the first
    2000 segments in a transaction will create trace segments if the
    transaction exceeds the trace theshold (4 x apdex_t).

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_max_trace_segments($txn_id, 5000);

 set_transaction_category

    Sets the transaction's category name (.e.g Uri in
    "WebTransaction/Uri/<txn_name>").

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_category($txn_id, 'Custom');

 set_transaction_type_web

    Sets the transaction type to WebTransaction. This will automatically
    change the category to Uri.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_type_web($txn_id);

 set_transaction_type_other

    Sets the transaction type to OtherTransaction. This will automatically
    change the category to Custom.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->set_transaction_type_other($txn_id);

 add_transaction_attribute

    Sets a transaction attribute. Up to the first 50 attributes added are
    sent with each transaction.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->add_transaction_attribute($txn_id, 'User-Agent', 'Mozilla/5.0 ...');

 notice_transaction_error

    Identify an error that occurred during the transaction. The first
    identified error is sent with each transaction.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->notice_transaction_error(
            $txn_id,
            'Runtime error',
            'Illegal division by zero',
            "Illegal division by zero at div0.pl line 4.\nmain::run() called at div0.pl line7",
            "\n",
        );

 end_transaction

    Identify the end of a transaction.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->end_transaction($txn_id);

 record_metric

    Record a custom metric.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->record_metric('cache_miss_timing', 0.333333);

 record_cpu_usage

    Record CPU user time in seconds and as a percentage of CPU capacity.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->record_cpu_usage(2.1, 0.85);

 record_memory_usage

    Record the current amount of memory (in megabytes) being used.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->record_memory_usage(745);

 begin_generic_segment

    Identify the beginning of a segment that performs a generic operation.
    This type of segment does not create metrics, but can show up in a
    transaction trace if a transaction is slow enough.

    Returns the segment's ID on success, else negative warning code or
    error code.

    Example:

        my $seg_id = $agent->begin_generic_segment($txn_id, undef, 'Parse zip codes');

 begin_datastore_segment

    Identify the beginning of a segment that performs a database operation.
    This uses the default sql_obfuscator that strips the SQL string
    literals and numeric sequences, replacing them with the ? character.

    Returns the segment's ID on success, else negative warning code or
    error code.

    Example:

        my $seg_id = $agent->begin_datastore_segment(
            $txn_id,
            undef,
            'users',
            'selecting user',
            'SELECT * FROM users WHERE id=?',
            'get_user_account',
        );

 begin_external_segment

    Identify the beginning of a segment that performs an external service.

    Returns the segment's ID on success, else negative warning code or
    error code.

    Example:

        my $seg_id = $agent->begin_external_segment(
            $txn_id,
            undef,
            'http://api.stripe.com/v1',
            'tokenize credit card',
        );

 end_segment

    Identify the end of a segment.

    Returns 0 on success, else negative warning code or error code.

    Example:

        my $err_id = $agent->end_segment($txn_id, $seg_id);

 get_license_key

    Returns the license key that the agent has loaded. This is useful for
    diagnostic purposes.

    Example:

        my $license_key = $agent->get_license_key;

 get_app_name

    Returns the application name that the agent has loaded. This is useful
    for diagnostic purposes.

    Example:

        my $app_name = $agent->get_app_name;

 get_app_language

    Returns the application language that the agent has loaded. This is
    useful for diagnostic purposes.

    Example:

        my $app_language = $agent->get_app_language;

 get_app_language_version

    Returns the application language's version that the agent has loaded.
    This is useful for diagnostic purposes.

    Example:

        my $app_language_version = $agent->get_app_language_version;

BUGS

    Please report any bugs or feature requests on the bugtracker website
    https://github.com/aanari/NewRelic-Agent/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR

    Ali Anari <ali@anari.me>

CONTRIBUTORS

      * Slobodan Mišković <slobodan@miskovic.ca>

      * Tatsuhiko Miyagawa <miyagawa@bulknews.net>

      * Tim Bunce <tim@tigerlms.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2015 by Ali Anari.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2015-03-09 00:00:00 +0000 to present.
==================================================

-------------------------------------------
version 0.0532 at 2016-03-08 03:48:11 +0000
-------------------------------------------

  Change: 7ceefb338ef975f797f56a5cee9b10a11e149840
  Author: Tatsuhiko Miyagawa <miyagawa@bulknews.net>
  Date : 2016-03-07 18:27:57 +0000

    bump 

  Change: 3897d1cb9f4919e7288df78fa2b72dbb19aa20a0
  Author: Tatsuhiko Miyagawa <miyagawa@bulknews.net>
  Date : 2016-03-07 18:27:09 +0000

    Move XSpp to configure requires 

  Change: b5965c886179a12a419b369e17df7e29378a793e
  Author: Tatsuhiko Miyagawa <miyagawa@bulknews.net>
  Date : 2016-03-04 16:23:06 +0000

    add MetaJSON 

  Change: 4bb12c1365d0cdfa181c71011f09af43e7f66cb6
  Author: Tatsuhiko Miyagawa <miyagawa@bulknews.net>
  Date : 2016-03-04 16:19:08 +0000

    add CommitBuild 

  Change: 08a6a8a48d4fe4af71ab27f1d789712aaee3e7c6
  Author: Tatsuhiko Miyagawa <miyagawa@bulknews.net>
  Date : 2016-03-04 16:11:13 +0000

    band-aid fix to require ParseXS in configure 

  Change: be8ccd57046935a283fc2e726324d006a85e1ba8
  Author: Ali Anari <ali@anari.me>
  Date : 2015-09-03 08:06:35 +0000

    Bumped version 

  Change: 4b6f93857e5d6d4aff1ba0e65abdbefbe8a1ef6a
  Author: Ali Anari <ali@anari.me>
  Date : 2015-09-03 08:00:50 +0000

    Merge pull request #4 from timbunce/remove-method-signatures

    Remove use of Method::Signatures, add arg checking (saves >35MB) 

  Change: ce5e310dc5ea06354eb2e8038ce171d208c18666
  Author: Ali Anari <ali@anari.me>
  Date : 2015-09-03 08:00:20 +0000

    Merge pull request #3 from timbunce/fix-01-basic-perl-version

    Fix t/01-basic.t to not assume/require 5.014002 

  Change: af5ea627c1b70d300ea18bbd0be44286834680f1
  Author: Ali Anari <ali@anari.me>
  Date : 2015-09-03 07:59:56 +0000

    Merge pull request #1 from slobo/patch-1

    Fix test relying being run on per 5.14.2 

  Change: 4cf70d9263220a17ac5c372f1629db69fa506e21
  Author: Tim Bunce <tim@tigerlms.com>
  Date : 2015-09-02 13:48:41 +0000

    Remove use of Method::Signatures, add arg checking (saves >35MB on my
    system) 

  Change: 1b4fa0560a9e7557d1a87e0806b68315888f979d
  Author: Tim Bunce <tim@tigerlms.com>
  Date : 2015-09-02 13:38:15 +0000

    Fix t/01-basic.t to not assume/require 5.014002 

  Change: 54bc373216f9df2e0be6fe8b47b8919b2ade43bd
  Author: Slobodan MiÅ¡koviÄ <slobodan@miskovic.ca>
  Date : 2015-07-05 09:45:12 +0000

    Fix test relying being run on per 5.14.2

    This made the build fail on anything but. 

-------------------------------------------
version 0.0510 at 2015-05-12 19:09:22 +0000
-------------------------------------------

  Change: dae25bb8b19169c8ad362aa61f73fcce15be2969
  Author: Ali Anari <ali@anari.me>
  Date : 2015-05-12 19:09:22 +0000

    Fixed absolute vs. relative path issue 

-------------------------------------------
version 0.0503 at 2015-05-07 14:40:10 +0000
-------------------------------------------

  Change: bd7234b1323c27e1b2b7a7c3cdeb06625053f31c
  Author: Ali Anari <ali@anari.me>
  Date : 2015-05-07 14:40:10 +0000

    Fixed abstract 

  Change: 1e0e00c297357440edb2bf9f99ee55d783113626
  Author: Ali Anari <ali@anari.me>
  Date : 2015-05-07 14:33:58 +0000

    Added README file 

-------------------------------------------
version 0.0501 at 2015-05-07 14:15:32 +0000
-------------------------------------------

  Change: e46bc306ddbc76e961c07d7906564036145bddce
  Author: Ali Anari <ali@anari.me>
  Date : 2015-05-07 14:15:32 +0000

    Added CHANGES file 

-------------------------------------------
version 0.0500 at 2015-04-30 21:58:15 +0000
-------------------------------------------

  Change: ff30eb072be8f4c3557e5ea1d3c2ae44df03479d
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-30 21:58:15 +0000

    Fixed destructor, embed_collector, and init 

-------------------------------------------
version 0.0420 at 2015-04-25 15:58:15 +0000
-------------------------------------------

  Change: c26f40d836304a86c5666cad4ec416015d36b1a7
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-25 15:58:15 +0000

    Fixed manifest .so issue 

-------------------------------------------
version 0.0410 at 2015-04-25 00:36:19 +0000
-------------------------------------------

  Change: 5561629e473e763588d029b0841b289a8de2c145
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-25 00:36:19 +0000

    Fixed relative to absolute paths 

  Change: 1b11ec85066f8830b648010090669057a7c516e9
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-22 20:01:16 +0000

    update travis config 

-----------------------------------------
version 0.04 at 2015-04-23 01:43:21 +0000
-----------------------------------------

  Change: 108ebe7d6b7567332fdb428bee79168c7c88b8bb
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-23 01:43:21 +0000

    travis badge 

  Change: af1309e52453e2e1e92315f3c8b84b689148eaf2
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-23 01:41:36 +0000

    Added TravisCI 

  Change: 443c0a54a60de3fa0d3febed81323f14ba83a6f6
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-23 01:38:04 +0000

    Added POD and dist plugins 

  Change: 47afd9f3d4b0b624529468f7f01c0b32b1729bb9
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-23 00:21:38 +0000

    Implemented rest of methods, bumped version 

  Change: 0477675a79f09bb6076a2bff1bebc26eca48837d
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-22 23:35:48 +0000

    Implemented transaction methods 

  Change: 3ed212f11225696ed75a3fa3d9aeff9bf9cba764
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-22 23:11:59 +0000

    Bumped version 

  Change: 6b84791e8535769298e195cc70d5642a9aaf3da0
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-22 23:11:42 +0000

    Lots of methods, fixed linker 

  Change: e7e7200e29461bc7a831a332e609d11afff1c2c4
  Author: Ali Anari <ali@anari.me>
  Date : 2015-04-22 22:07:31 +0000


Makefile.PL
# This Makefile.PL for NewRelic-Agent was generated by
# inc::AgentMakeMaker <self>
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.35.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl Agent for NewRelic APM",
  "AUTHOR" => "Ali Anari <ali\@anari.me>",
  "CC" => "g++",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "ExtUtils::ParseXS" => "3.30",
    "ExtUtils::XSpp" => "0.18"
  },
  "DISTNAME" => "NewRelic-Agent",
  "INC" => "-I. -Isdk/include",
  "LD" => "\$(CC)",
  "LIBS" => "-Lsdk/lib -lnewrelic-common -Lsdk/lib -lnewrelic-collector-client -Lsdk/lib -lnewrelic-transaction",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "NewRelic::Agent",
  "OBJECT" => "\$(O_FILES)",
  "PMLIBDIRS" => [
    "lib",
    "\$(BASEEXT)",
    "sdk/lib"
  ],
  "PREREQ_PM" => {
    "Carp" => 0,
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0,
    "blib" => "1.01"
  },
  "VERSION" => "0.0532",
  "XSOPT" => "-C++ -hiertype",
  "depend" => {
    "WithAgent.c" => "NewRelic-Agent.xsp"
  },
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Carp" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => 0,
  "XSLoader" => 0,
  "blib" => "1.01",
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/NewRelic/Agent.pm
package NewRelic::Agent;
use strict;
use warnings;

our $VERSION = '0.0532';

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub new {
    my ($self, %args) = @_;

    my $license_key          = delete $args{license_key}
                            || $ENV{NEWRELIC_LICENSE_KEY}
                            || '';
    my $app_name             = delete $args{app_name}
                            || $ENV{NEWRELIC_APP_NAME}
                            || 'AppName';
    my $app_language         = delete $args{app_language}
                            || $ENV{NEWRELIC_APP_LANGUAGE}
                            || 'perl';
    my $app_language_version = delete $args{app_language_version}
                            || $ENV{NEWRELIC_APP_LANGUAGE_VERSION}
                            || $];

    if (%args) {
        require Carp;
        Carp::croak("Invalid arguments: @{[ keys %args ]}");
    }

    return $self->_new($license_key, $app_name, $app_language, $app_language_version);
}

1;

# ABSTRACT: Perl Agent for NewRelic APM

__END__

=pod

=encoding UTF-8

=head1 NAME

NewRelic::Agent - Perl Agent for NewRelic APM

=head1 VERSION

version 0.0532

=head1 SYNOPSIS

    use NewRelic::Agent;

    my $agent = NewRelic:Agent->new(
        license_key => 'abc123',
        app_name    => 'REST API',
    );

    $agent->embed_collector;
    $agent->init;
    my $txn_id = $agent->begin_transaction;
    ...
    my $err_id = $agent->end_transaction($txn_id);

=head1 DESCRIPTION

This module provides bindings for the L<NewRelic|https://docs.newrelic.com/docs/agents/agent-sdk/getting-started/new-relic-agent-sdk> Agent SDK.

=for markdown [![Build Status](https://travis-ci.org/aanari/NewRelic-Agent.svg?branch=master)](https://travis-ci.org/aanari/NewRelic-Agent)

=head1 METHODS

=head2 new

Instantiates a new NewRelic::Agent client object.

    my $agent = NewRelic::Agent->new(
        license_key          => $license_key,
        app_name             => $app_name,
        app_language         => $app_language,         #optional
        app_language_version => $app_language_version, #optional
    );

B<Parameters>

=over 4

=item - C<license_key>

A valid NewRelic license key for your account.

This value is also automatically sourced from the C<NEWRELIC_LICENSE_KEY> environment variable.

=item - C<app_name>

The name of your application.

This value is also automatically sourced from the C<NEWRELIC_APP_NAME> environment variable.

=item - C<app_language>

The language that your application is written in.

This value defaults to C<perl>, and can also be automatically sourced from the C<NEWRELIC_APP_LANGUAGE> environment variable.

=item - C<app_language_version>

The version of the language that your application is written in.

This value defaults to your perl version, and can also be automatically sourced from the C<NEWRELIC_APP_LANGUAGE_VERSION> environment variable.

=back

=head2 embed_collector

Embeds the collector agent for harvesting NewRelic data. This should be called before C<init>, if the agent is being used in Embedded mode and not Daemon mode.

B<Example:>

    $agent->embed_collector;

=head2 init

Initialize the connection to NewRelic.

B<Example:>

    $agent->init;

=head2 begin_transaction

Identifies the beginning of a transaction, which is a timed operation consisting of multiple segments. By default, transaction type is set to C<WebTransaction> and transaction category is set to C<Uri>.

Returns the transaction's ID on success, else negative warning code or error code.

B<Example:>

    my $txn_id = $agent->begin_transaction;

=head2 set_transaction_name

Sets the transaction's name.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_name($txn_id, 'Create Account');

=head2 set_transaction_request_url

Sets the transaction's request url. The query part of the url is automatically stripped from the url.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_request_url($txn_id, 'api.myapp.com/users/123');

=head2 set_transaction_max_trace_segments

Sets the maximum number of trace segments allowed in a transaction trace. By default, the maximum is set to C<2000>, which means the first 2000 segments in a transaction will create trace segments if the transaction exceeds the trace theshold (4 x apdex_t).

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_max_trace_segments($txn_id, 5000);

=head2 set_transaction_category

Sets the transaction's category name (.e.g C<Uri> in "WebTransaction/Uri/<txn_name>").

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_category($txn_id, 'Custom');

=head2 set_transaction_type_web

Sets the transaction type to C<WebTransaction>. This will automatically change the category to C<Uri>.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_type_web($txn_id);

=head2 set_transaction_type_other

Sets the transaction type to C<OtherTransaction>. This will automatically change the category to C<Custom>.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->set_transaction_type_other($txn_id);

=head2 add_transaction_attribute

Sets a transaction attribute. Up to the first 50 attributes added are sent with each transaction.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->add_transaction_attribute($txn_id, 'User-Agent', 'Mozilla/5.0 ...');

=head2 notice_transaction_error

Identify an error that occurred during the transaction. The first identified error is sent with each transaction.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->notice_transaction_error(
        $txn_id,
        'Runtime error',
        'Illegal division by zero',
        "Illegal division by zero at div0.pl line 4.\nmain::run() called at div0.pl line7",
        "\n",
    );

=head2 end_transaction

Identify the end of a transaction.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->end_transaction($txn_id);

=head2 record_metric

Record a custom metric.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->record_metric('cache_miss_timing', 0.333333);

=head2 record_cpu_usage

Record CPU user time in seconds and as a percentage of CPU capacity.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->record_cpu_usage(2.1, 0.85);

=head2 record_memory_usage

Record the current amount of memory (in megabytes) being used.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->record_memory_usage(745);

=head2 begin_generic_segment

Identify the beginning of a segment that performs a generic operation. This type of segment does not create metrics, but can show up in a transaction trace if a transaction is slow enough.

Returns the segment's ID on success, else negative warning code or error code.

B<Example:>

    my $seg_id = $agent->begin_generic_segment($txn_id, undef, 'Parse zip codes');

=head2 begin_datastore_segment

Identify the beginning of a segment that performs a database operation. This uses the default sql_obfuscator that strips the SQL string literals and numeric sequences, replacing them with the C<?> character.

Returns the segment's ID on success, else negative warning code or error code.

B<Example:>

    my $seg_id = $agent->begin_datastore_segment(
        $txn_id,
        undef,
        'users',
        'selecting user',
        'SELECT * FROM users WHERE id=?',
        'get_user_account',
    );

=head2 begin_external_segment

Identify the beginning of a segment that performs an external service.

Returns the segment's ID on success, else negative warning code or error code.

B<Example:>

    my $seg_id = $agent->begin_external_segment(
        $txn_id,
        undef,
        'http://api.stripe.com/v1',
        'tokenize credit card',
    );

=head2 end_segment

Identify the end of a segment.

Returns C<0> on success, else negative warning code or error code.

B<Example:>

    my $err_id = $agent->end_segment($txn_id, $seg_id);

=head2 get_license_key

Returns the license key that the agent has loaded. This is useful for diagnostic purposes.

B<Example:>

    my $license_key = $agent->get_license_key;

=head2 get_app_name

Returns the application name that the agent has loaded. This is useful for diagnostic purposes.

B<Example:>

    my $app_name = $agent->get_app_name;

=head2 get_app_language

Returns the application language that the agent has loaded. This is useful for diagnostic purposes.

B<Example:>

    my $app_language = $agent->get_app_language;

=head2 get_app_language_version

Returns the application language's version that the agent has loaded. This is useful for diagnostic purposes.

B<Example:>

    my $app_language_version = $agent->get_app_language_version;

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/aanari/NewRelic-Agent/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Ali Anari <ali@anari.me>

=head1 CONTRIBUTORS

=for stopwords Slobodan Mišković Tatsuhiko Miyagawa Tim Bunce

=over 4

=item *

Slobodan Mišković <slobodan@miskovic.ca>

=item *

Tatsuhiko Miyagawa <miyagawa@bulknews.net>

=item *

Tim Bunce <tim@tigerlms.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Ali Anari.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
